research/eccv2024/model_evaluation/fgrained_model_evaluation.py

```diff
@@ -0,0 +1,264 @@
+
+    for prediction in sp_pred:
+        # Get family and genus name for every species prediction
+        sp_key, conf = prediction[0], round(float(prediction[1]), 3)
+        genus = gbif_taxonomy_hierarchy[sp_key][0]
+        family = gbif_taxonomy_hierarchy[sp_key][1]
+
+        # Genus accuracy calculation
+        if genus not in genus_pred.keys(): genus_pred[genus] = conf
+        else: genus_pred[genus] += conf
+
+        # Family accuracy calculation
+        if family not in family_pred.keys(): family_pred[family] = conf
+        else: family_pred[family] += conf
+
+    # Sort the prediction in decreasing order of confidence
+    genus_pred_sorted = [[taxa, round(conf, 3)] for taxa, conf in sorted(list(genus_pred.items()), key=lambda item: item[1], reverse=True)]
+    family_pred_sorted = [[taxa, round(conf, 3)] for taxa, conf in sorted(list(family_pred.items()), key=lambda item: item[1], reverse=True)]
+
+    return genus_pred_sorted, family_pred_sorted
+
+
+
+def get_higher_taxon_gt(label: str, rank: str, taxonomy_map: pd.DataFrame):
+    """Get higher taxon for a ground truth label"""
+    
+    if rank == "SPECIES":
+        try: 
+            query = taxonomy_map.loc[taxonomy_map["name"] == label]
+            return query["GENUS"].values[0], query["FAMILY"].values[0]
+        except: 
+            print(f"{label} of rank {rank} not found in the taxonomy database.")
+
+    else:
+        try: 
+            query = taxonomy_map.loc[taxonomy_map["name"] == label]
+            return query["FAMILY"].values[0]
+        except: 
+            print(f"{label} of rank {rank} not found in the taxonomy database.")
+
+
+
+def accuracy_versus_confidence(
+    gt_label: str, 
+    pred_label: list[list[str, float]], 
+    accuracy_vs_conf_df: pd.DataFrame
+):
+    """Calculate accuracy alongside model prediction confidence"""
+
+    top1_class, top1_acc = pred_label[0][0], math.floor(pred_label[0][1]*100)
+    thresh = 0 # for cases with confidence < 10%
+
+    for thresh in range(10, top1_acc, 10):  
+        # Add the counts upto the threshold  
+        accuracy_vs_conf_df.loc["conf-total"]["conf-"+str(thresh)] += 1
+
+        # Add only correct counts
+        if top1_class == gt_label: 
+            accuracy_vs_conf_df.loc["correct"]["conf-"+str(thresh)] += 1
+
+    # Add counts for upwards rejected thresholds
+    for upwards_thresh in range(thresh+10, 100, 10):
+        accuracy_vs_conf_df.loc["reject"]["conf-"+str(upwards_thresh)] += 1
+
+    return accuracy_vs_conf_df
+
+
+def fgrained_model_evaluation(
+    run_name: str,
+    artifact: str,
+    region: str,
+    model_type: str,
+    model_dir: str,
+    category_map: str,
+    insect_crops_dir: str,
+    sp_exclusion_list_file: str,
+    ami_traps_taxonomy_map_file: str,
+    ami_gbif_taxonomy_map_file: str,
+    gbif_taxonomy_hierarchy_file: str,
+    save_acccuracy_vs_conf: bool = True
+):
+    """Main function for fine-grained model evaluation"""
+
```

Comment:
Similar comment about both orchestrating and processing in the same function

Reply:
Fixed (mostly) in ce73ced.